### ISSUE TITLE
CTCLoss is slow on TPU

### ISSUE THREAD

**Issue by discort** [pytorch/xla#2844]
## 🐛 Bug

Backprop of [CTCLoss](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html) is calculated slowly. 
I got the [metric report](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#get-a-metrics-report) and found `aten::_ctc_loss` and `aten::_ctc_loss_backward`:
```
Metric: CompileTime
  TotalSamples: 5
  Accumulator: 10m24s127ms829.269us
  ValueRate: 981ms105.997us / second
  Rate: 0.00785983 / second
  Percentiles: 1%=051ms800.759us; 5%=051ms800.759us; 10%=051ms800.759us; 20%=01m07s843ms564.327us; 50%=01m07s293ms066.106us; 80%=07m02s560ms545.030us; 90%=07m02s560ms545.030us; 95%=07m02s560ms545.030us; 99%=07m02s560ms545.030us
Metric: DeviceLockWait
  TotalSamples: 5
  Accumulator: 023.429us
  ValueRate: 000.110us / second
  Rate: 0.0234415 / second
  Percentiles: 1%=002.952us; 5%=002.952us; 10%=002.952us; 20%=003.956us; 50%=005.385us; 80%=005.595us; 90%=005.595us; 95%=005.595us; 99%=005.595us
Metric: ExecuteTime
  TotalSamples: 4
  Accumulator: 01s069ms641.694us
  ValueRate: 005ms024.360us / second
  Rate: 0.0188065 / second
  Percentiles: 1%=006ms770.412us; 5%=006ms770.412us; 10%=006ms770.412us; 20%=006ms770.412us; 50%=368ms423.482us; 80%=372ms988.197us; 90%=372ms988.197us; 95%=372ms988.197us; 99%=372ms988.197us
Metric: InboundData
  TotalSamples: 6
  Accumulator: 36.65MB
  ValueRate: 180.36KB / second
  Rate: 0.0288332 / second
  Percentiles: 1%=1.00KB; 5%=1.00KB; 10%=1.00KB; 20%=1.00KB; 50%=11.94MB; 80%=12.22MB; 90%=12.22MB; 95%=12.22MB; 99%=12.22MB
Metric: InputOutputAliasCount
  TotalSamples: 4
  Accumulator: 79.00
  ValueRate: 0.37 / second
  Rate: 0.0186604 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=18.00; 80%=57.00; 90%=57.00; 95%=57.00; 99%=57.00
Metric: IrValueTensorToXlaData
  TotalSamples: 21
  Accumulator: 444ms043.225us
  ValueRate: 003ms138.219us / second
  Rate: 0.148415 / second
  Percentiles: 1%=001ms280.656us; 5%=001ms329.654us; 10%=001ms367.987us; 20%=001ms462.492us; 50%=006ms633.329us; 80%=043ms382.522us; 90%=057ms569.944us; 95%=057ms302.788us; 99%=067ms227.154us
Metric: OutboundData
  TotalSamples: 27
  Accumulator: 655.15MB
  ValueRate: 4.55MB / second
  Rate: 0.187391 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=512.00B; 50%=8.00KB; 80%=16.00MB; 90%=16.00MB; 95%=264.76MB; 99%=264.90MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 4
  Accumulator: 007ms050.146us
  ValueRate: 033.137us / second
  Rate: 0.0188005 / second
  Percentiles: 1%=001ms188.942us; 5%=001ms188.942us; 10%=001ms188.942us; 20%=001ms188.942us; 50%=001ms410.698us; 80%=003ms214.407us; 90%=003ms214.407us; 95%=003ms214.407us; 99%=003ms214.407us
Metric: TensorsGraphSize
  TotalSamples: 5
  Accumulator: 152398.00
  ValueRate: 239.53 / second
  Rate: 0.00785878 / second
  Percentiles: 1%=90.00; 5%=90.00; 10%=90.00; 20%=18651.00; 50%=18652.00; 80%=96340.00; 90%=96340.00; 95%=96340.00; 99%=96340.00
Metric: TransferFromServerTime
  TotalSamples: 6
  Accumulator: 223ms854.071us
  ValueRate: 001ms070.933us / second
  Rate: 0.0288332 / second
  Percentiles: 1%=001ms467.111us; 5%=001ms467.111us; 10%=001ms467.111us; 20%=005ms699.882us; 50%=064ms495.619us; 80%=070ms709.854us; 90%=077ms929.425us; 95%=077ms929.425us; 99%=077ms929.425us
Metric: TransferToServerTime
  TotalSamples: 27
  Accumulator: 02s282ms299.203us
  ValueRate: 016ms841.027us / second
  Rate: 0.187402 / second
  Percentiles: 1%=001ms259.324us; 5%=001ms270.733us; 10%=001ms319.534us; 20%=001ms450.514us; 50%=006ms624.463us; 80%=051ms290.135us; 90%=067ms216.359us; 95%=912ms907.359us; 99%=915ms051.091us
Metric: TransferToServerTransformTime
  TotalSamples: 27
  Accumulator: 196ms041.358us
  ValueRate: 001ms360.609us / second
  Rate: 0.187391 / second
  Percentiles: 1%=102.625us; 5%=119.517us; 10%=142.757us; 20%=173.034us; 50%=389.501us; 80%=004ms000.432us; 90%=011ms996.010us; 95%=067ms565.028us; 99%=082ms978.278us
Counter: CreateCompileHandles
  Value: 5
Counter: CreateDataHandles
  Value: 80
Counter: CreateXlaTensor
  Value: 48256
Counter: DestroyDataHandles
  Value: 23
Counter: DestroyXlaTensor
  Value: 48171
Counter: DeviceDataCacheMiss
  Value: 4
Counter: MarkStep
  Value: 2
Counter: ReleaseDataHandles
  Value: 26
Counter: UncachedCompile
  Value: 5
Counter: XRTAllocateFromTensor_Empty
  Value: 31
Counter: XrtCompile_Empty
  Value: 32
Counter: XrtExecuteChained_Empty
  Value: 32
Counter: XrtExecute_Empty
  Value: 32
Counter: XrtMemoryInfo_Empty
  Value: 32
Counter: XrtRead_Empty
  Value: 32
Counter: XrtReleaseAllocationHandle_Empty
  Value: 32
Counter: XrtReleaseCompileHandle_Empty
  Value: 32
Counter: XrtSessionCount
  Value: 3
Counter: XrtSubTuple_Empty
  Value: 32
Counter: aten::_ctc_loss
  Value: 1
Counter: aten::_ctc_loss_backward
  Value: 1
Counter: xla::_copy_from
  Value: 29
Counter: xla::_log_softmax
  Value: 1
Counter: xla::_log_softmax_backward_data
  Value: 1
Counter: xla::_unsafe_view
  Value: 5
Counter: xla::add
  Value: 9266
Counter: xla::add_
  Value: 1709
Counter: xla::addmm
  Value: 3372
Counter: xla::as_strided
  Value: 19
Counter: xla::cat
  Value: 1
Counter: xla::clamp_min
  Value: 1
Counter: xla::clone
  Value: 18
Counter: xla::div
  Value: 3
Counter: xla::empty
  Value: 27
Counter: xla::empty_strided
  Value: 19
Counter: xla::expand
  Value: 1
Counter: xla::fill_
  Value: 1
Counter: xla::hardtanh
  Value: 4
Counter: xla::hardtanh_backward
  Value: 4
Counter: xla::mean
  Value: 1
Counter: xla::mm
  Value: 6756
Counter: xla::mul
  Value: 1
Counter: xla::relu
  Value: 1690
Counter: xla::slice
  Value: 14
Counter: xla::squeeze
  Value: 1
Counter: xla::stack
  Value: 4
Counter: xla::sum
  Value: 3377
Counter: xla::t
  Value: 13505
Counter: xla::threshold_backward
  Value: 1690
Counter: xla::transpose
  Value: 3
Counter: xla::unbind
  Value: 4
Counter: xla::unsqueeze
  Value: 1
Counter: xla::view
  Value: 3396
Counter: xla::zero_
  Value: 7
Metric: XrtAllocateFromTensor
  TotalSamples: 1319
  Accumulator: 39s463ms641.542us
  Mean: 036ms450.058us
  StdDev: 244ms634.889us
  Rate: 0.906315 / second
  Percentiles: 25%=482.074us; 50%=002ms610.773us; 80%=006ms264.427us; 90%=021ms526.056us; 95%=024ms354.515us; 99%=02s930ms384.947us
Metric: XrtCompile
  TotalSamples: 174
  Accumulator: 01h04m16s613ms925.298us
  Mean: 22s159ms694.973us
  StdDev: 52s739ms482.196us
  Rate: 0.105229 / second
  Percentiles: 25%=026ms486.182us; 50%=038ms657.839us; 80%=18s622ms441.773us; 90%=02m49s194ms114.553us; 95%=02m21s149ms854.395us; 99%=03m35s397ms155.654us
Metric: XrtExecute
  TotalSamples: 199
  Accumulator: 26m14s659ms615.185us
  Mean: 08s908ms832.237us
  StdDev: 29s427ms294.247us
  Rate: 0.161777 / second
  Percentiles: 25%=052ms198.087us; 50%=063ms870.385us; 80%=242ms157.991us; 90%=07s229ms246.183us; 95%=02m31s235ms179.971us; 99%=03m46s636ms115.495us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles:
Metric: XrtReadLiteral
  TotalSamples: 458
  Accumulator: 954ms041.328us
  Mean: 002ms083.060us
  StdDev: 002ms882.552us
  Rate: 0.375839 / second
  Percentiles: 25%=700.355us; 50%=002ms355.985us; 80%=003ms865.589us; 90%=003ms059.056us; 95%=003ms377.271us; 99%=006ms102.693us
Metric: XrtReleaseAllocation
  TotalSamples: 548
  Accumulator: 041ms686.423us
  Mean: 074.245us
  StdDev: 084.164us
  Rate: 0.33137 / second
  Percentiles: 25%=029.500us; 50%=050.734us; 80%=092.310us; 90%=141.818us; 95%=218.368us; 99%=448.588us
```

## To Reproduce

1. Run the collab [file](https://colab.research.google.com/drive/1Smh0khioEIieeM_iZj58nwBNVLOxAwU5?usp=sharing)

## Expected behavior

I'm expecting that `_ctc_loss` and `_ctc_loss_backward` will be calculated on TPU like that they could be calculated on GPU.

## Environment

 - Reproducible on XLA backend [CPU/TPU]:
 - torch_xla version: `cloud-tpu-client==0.10`

**Comment by discort** [comment]
Just have found similar https://github.com/pytorch/xla/issues/2163#issuecomment-639486271 

@davidel 
just wondering are you planning to add support of CTCLoss in xla into your queue?

**Comment by JackCaoG** [comment]
We already have https://github.com/pytorch/xla/issues/2399 open, I will close this one. We will try to lower this one soon (~2 months)
